### PR TITLE
PasswordReset fix for #1374 and #947 

### LINKED ISSUE #947: in User.resetPassword callback called without error if user not founded
In line https://github.com/strongloop/loopback/blob/master/common/models/user.js#L445
called callback without error about not founded user by email.

```
// on existed email - PASSED
app.models.User.resetPassword({email: 'existed@mail.com'}, function (err) {
 assert(err === null);
});

// not existed email - FAILED
app.models.User.resetPassword({email: 'not-existed@mail.com'}, function (err) {
 assert(err);
});
```

Sorry for my poor English.

### LINKED ISSUE #1374: User.resetPassword does not return error when email is not found
There's a miss in the User.resetPassword code where the callback does not clarify when a user isn't found for the supplied email. 

See my annotation in the code below and a suggested fix:
https://github.com/strongloop/loopback/blob/master/common/models/user.js#L479

```js
User.resetPassword = function(options, cb) {

...

    UserModel.findOne({ where: {email: options.email} }, function(err, user) {
      if (err) {
        cb(err);
      } else if (user) {
        // create a short lived access token for temp login to change password
        // TODO(ritch) - eventually this should only allow password change
        user.accessTokens.create({ttl: ttl}, function(err, accessToken) {
          if (err) {
            cb(err);
          } else {
            cb();
            UserModel.emit('resetPasswordRequest', {
              email: options.email,
              accessToken: accessToken,
              user: user
            });
          }
        });
      } else {
        cb(); // <<<<<<<<<<----- This does not clearly report that the user record was not found
      }
    });
...
};
```

Suggested fix:
```

UserModel.findOne({ where: {email: options.email} }, function(err, user) {
      if (err || !user) {
        return cb(err ||  'Email not found');
      }
      user.accessTokens.create({ttl: ttl}, function(err, accessToken) {
          if (err) {
            cb(err);
          } else {
            cb();
            UserModel.emit('resetPasswordRequest', {
              email: options.email,
              accessToken: accessToken,
              user: user
            });
          }
      });

});

```

Thanks!

### PR DESCRIPTION
**Pull request by simoami**
@superkhau 
Bummer, I saw the other PR only after finishing this one. But you might like the code re-structure. I've also added a unit test.

Fix #1374 and fix #947

### CODE REVIEW

**Comment by slnode** [comment]
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

File: common/models/user.js [hunk 1]
```diff
@@ -506,35 +506,39 @@ module.exports = function(User) {
...
-              });
-            }
-          });
-        } else {
-          cb();
-        }
-      });
-    } else {
+    if (typeof options.email !== 'string') {
       var err = new Error('email is required');
```
Comment (superkhau) [inline common/models/user.js]: "Email is required" to match "Email not found" below.

**Comment by superkhau** [comment]
Other than minor comments, LGTM. @raymondfeng for the final OK.

**Comment by bajtos** [comment]
@slnode ok to test

**Comment by bajtos** [comment]
I have fixed the error message myself and cleaned up the git history.

The patch was landed, thank you for the contribution!

**Comment by simoami** [comment]
:thumbsup:

**Pull request closed without merging**